In [147]:
%matplotlib inline
import pandas as pd

In [148]:
cast = pd.read_csv('C:/Users/Akash/Downloads/cast.csv')
cast.head()

title  year      name   type                character     n
0        Closet Monster  2015  Buffy #1  actor                  Buffy 4   NaN
1       Suuri illusioni  1985    Homo $  actor                   Guests  22.0
2   Battle of the Sexes  2017   $hutter  actor          Bobby Riggs Fan  10.0
3  Secret in Their Eyes  2015   $hutter  actor          2002 Dodger Fan   NaN
4            Steve Jobs  2015   $hutter  actor  1988 Opera House Patron   NaN

In [149]:
cast.count()

title        3611212
year         3611212
name         3611212
type         3611212
character    3611212
n            2231970
dtype: int64

### How many movies are listed in the titles dataframe?

In [6]:
len(cast.title)

3611212

#### In movies there are 36,11,212 titles is listed in the dataframe.

### What are the earliest two films listed in the titles dataframe?

In [54]:
cast.sort_values(by='year').head(2)

title  year             name     type  \
2514470  Miss Jerry  1894  Blanche Bayliss  actress   
555929   Miss Jerry  1894   Chauncey Depew    actor   

                                                 character   n  
2514470               Miss Geraldine Holbrook (Miss Jerry) NaN  
555929   Himself - the Director of the New York Central... NaN

#### In the data Frame earliest two films is Miss jerry and Miss Jerry

### How many movies have the title "Hamlet"?

In [29]:
len(cast[cast.title == "Hamlet"])

313

#### 313 moves has Title 'hamlet'

### How many movies are titled "North by Northwest"?

In [30]:
len(cast[cast.title == "North by Northwest"])

112

#### 112 moves has Title 'North by Northwest'

### When was the first movie titled "Hamlet" made?

In [56]:
cast[cast.title=="Hamlet"].year.min()

1910

#### In 1910 was the first movie titled "Hamlet"

### List all of the "Treasure Island" movies from earliest to most recent.

In [51]:
cast[cast.title=="Treasure Island"].sort_values(by='year')

title  year                  name     type  \
830114   Treasure Island  1918        Charles Gorman    actor   
840745   Treasure Island  1918             Joe Grant    actor   
909157   Treasure Island  1918          Edwin Harley    actor   
3298306  Treasure Island  1918      Violet Radcliffe  actress   
1715014  Treasure Island  1918            Lloyd Perl    actor   
...                  ...   ...                   ...      ...   
2420089  Treasure Island  1999          Caveh Zahedi    actor   
1637661  Treasure Island  1999         Nick Offerman    actor   
1375015  Treasure Island  1999           J.P. Manoux    actor   
1797264  Treasure Island  1999  Victor Raider-Wexler    actor   
307749   Treasure Island  1999          Bob Byington    actor   

                       character     n  
830114           Prologue Player   NaN  
840745   Undetermined child role   NaN  
909157           Prologue Player   NaN  
3298306         Long John Silver   3.0  
1715014                Black Dog   4.0  
...                          ...   ...  
2420089                   Harold   9.0  
1637661                   Samuel   2.0  
1375015           Officer Hughes  18.0  
1797264                 Samowitz  10.0  
307749                    Thomas  12.0  

[190 rows x 6 columns]

#### Eraliest one is 1918 and recent one is 1999 

### How many movies were made in the year 1950?

In [60]:
len(cast[cast.year==1950])

22006

#### In the year 1950 there was 22,006 movies made in the year

### How many movies were made in the year 1960?

In [61]:
len(cast[cast.year==1960])

18456

#### 18,456 movies made in the year of 1960. 

### How many movies were made from 1950 through 1959?

In [64]:
len(cast[(cast.year>=1950)&(cast.year<=1959)])

210986

#### Between 1950 to 1959 there 2,10,986 movies made.

### In what years has a movie titled "Batman" been released?

In [75]:
cast[cast.title=="Batman"]

title  year               name     type                   character  \
95672    Batman  1943       Frank Austin    actor                 Hotel Clerk   
96186    Batman  1943     William Austin    actor           Alfred Pennyworth   
114034   Batman  1989          Del Baker    actor                 Napier Hood   
119476   Batman  1989    Michael Balfour    actor                   Scientist   
149767   Batman  1989         David Baxt    actor            Dr. Thomas Wayne   
...         ...   ...                ...      ...                         ...   
2907449  Batman  1989        Sharon Holm  actress                  Mrs. Wayne   
3253550  Batman  1943  Shirley Patterson  actress                  Linda Page   
3343947  Batman  1989           Pam Rose  actress  Party Guest at Wayne Manor   
3345872  Batman  1989          Liza Ross  actress                 Tourist Mom   
3356826  Batman  1989        Rachel Ryan  actress                      Corpse   

            n  
95672     NaN  
96186     NaN  
114034   31.0  
119476   44.0  
149767   48.0  
...       ...  
2907449  49.0  
3253550   4.0  
3343947   NaN  
3345872  45.0  
3356826   NaN  

[120 rows x 6 columns]

#### From 1943 movie titled "Batman"

### How many roles were there in the movie "Inception"?

In [78]:
len(cast[cast.title=="Inception"])

77

#### 77 Roles are there in "Inception" Movie

### How many roles in the movie "Inception" are NOT ranked by an "n" value?

In [79]:
len(cast[(cast.title=="Inception")&(cast.n.isnull())])

26

#### 26 values are roles in the movie "inception" are NOt ranked by an "n" value.

### But how many roles in the movie "Inception" did receive an "n" value?

In [81]:
len(cast[(cast.title=="Inception") &(cast.n.notnull())])

51

#### 51 roles in the movies "Inception" did recived an "n" Value.

### Display the cast of "North by Northwest" in their correct "n"-value order, ignoring roles that did not earn a numeric "n" value.

In [84]:
cast[(cast.title=="North by Northwest")&(cast.n.notnull())].sort_values(by='n')

title  year                  name     type  \
840374   North by Northwest  1959            Cary Grant    actor   
3361396  North by Northwest  1959       Eva Marie Saint  actress   
1404704  North by Northwest  1959           James Mason    actor   
3024085  North by Northwest  1959   Jessie Royce Landis  actress   
343701   North by Northwest  1959        Leo G. Carroll    actor   
2923056  North by Northwest  1959  Josephine Hutchinson  actress   
1634552  North by Northwest  1959           Philip Ober    actor   
1229083  North by Northwest  1959         Martin Landau    actor   
2357939  North by Northwest  1959         Adam Williams    actor   
1746383  North by Northwest  1959          Edward Platt    actor   
642085   North by Northwest  1959     Robert Ellenstein    actor   
2212083  North by Northwest  1959          Les Tremayne    actor   
447893   North by Northwest  1959       Philip Coolidge    actor   
1454363  North by Northwest  1959         Patrick McVey    actor   
198496   North by Northwest  1959          Edward Binns    actor   
1334726  North by Northwest  1959             Ken Lynch    actor   

                           character     n  
840374            Roger O. Thornhill   1.0  
3361396                  Eve Kendall   2.0  
1404704              Phillip Vandamm   3.0  
3024085              Clara Thornhill   4.0  
343701                 The Professor   5.0  
2923056                Mrs. Townsend   6.0  
1634552              Lester Townsend   7.0  
1229083                      Leonard   8.0  
2357939                     Valerian   9.0  
1746383              Victor Larrabee  10.0  
642085                         Licht  11.0  
2212083                   Auctioneer  12.0  
447893                     Dr. Cross  13.0  
1454363               Sergeant Flamm  14.0  
198496                Captain Junket  15.0  
1334726  Charley - Chicago Policeman  16.0

### Display the entire cast, in "n"-order, of the 1972 film "Sleuth".

In [86]:
cast[(cast.title=="Sleuth")&(cast.year==1972)].sort_values(by='n')

title  year                name     type  \
1644328  Sleuth  1972    Laurence Olivier    actor   
314691   Sleuth  1972       Michael Caine    actor   
360365   Sleuth  1972      Alec Cawthorne    actor   
1412647  Sleuth  1972  John (II) Matthews    actor   
2619278  Sleuth  1972  Eve (III) Channing  actress   
1396572  Sleuth  1972        Teddy Martin    actor   

                          character    n  
1644328                 Andrew Wyke  1.0  
314691                  Milo Tindle  2.0  
360365            Inspector Doppler  3.0  
1412647  Detective Sergeant Tarrant  4.0  
2619278             Marguerite Wyke  5.0  
1396572      Police Constable Higgs  6.0

#### In cast data there is 6 time movie reales in the yaer 1972 film "Sleuth"

### Now display the entire cast, in "n"-order, of the 2007 version of "Sleuth".

In [87]:
cast[(cast.title=="Sleuth")&(cast.year==2007)].sort_values(by='n')

title  year                 name     type          character    n
314692   Sleuth  2007        Michael Caine    actor             Andrew  1.0
1246855  Sleuth  2007             Jude Law    actor               Milo  2.0
1741124  Sleuth  2007        Harold Pinter    actor        Man on T.V.  3.0
250297   Sleuth  2007      Kenneth Branagh    actor  Other Man on T.V.  NaN
360366   Sleuth  2007  Alec (II) Cawthorne    actor  Inspector Doppler  NaN
2619277  Sleuth  2007    Eve (II) Channing  actress    Marguerite Wyke  NaN
3223502  Sleuth  2007    Carmel O'Sullivan  actress             Maggie  NaN

#### In 2007 there is 7 version released film "Sleuth" 

### How many roles were credited in the silent 1921 version of Hamlet?

In [91]:
len(cast[(cast.title=="Hamlet")&(cast.year==1921)])

9

#### 9 Roles were credited in the silent 1921 version of Hamlet

### How many roles were credited in Branagh’s 1996 Hamlet?

In [92]:
len(cast[(cast.title=="Hamlet")&(cast.year==1996)])

54

#### 54 Roles were credited in the silent 1921 version of Hamlet

### How many "Hamlet" roles have been listed in all film credits through history?

In [96]:
len(cast[cast.character=="Hamlet"])

91

#### 91 roles had been listed in all film "Hamlet"

### How many people have played an "Ophelia"?

In [98]:
len(cast[cast.character=="Ophelia"])

111

#### 111 peopler have played an "Ophelia"

### How many people have played a role called "The Dude"?

In [99]:
len(cast[cast.character=="The Dude"])

18

#### 18 peopler have played an "The Dude"

### How many people have played a role called "The Stranger"?

In [101]:
len(cast[cast.character=="The Stranger"])

208

#### 208 peope=le have played a role called "The Stranger"

### How many roles has Sidney Poitier played throughout his career?

In [103]:
len(cast[cast.name=="Sidney Poitier"])

43

#### 43 roles had played by Sidney Poitier

### How many roles has Judi Dench played?

In [104]:
len(cast[cast.name=="Judi Dench"])

55

#### 55 roles had played by Judi Dench

### List the supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year.

In [114]:
cast[(cast.name=="Cary Grant")&(cast.year//10==194)&(cast.n==2)].sort_values(by='year')

title  year        name   type    character    n
840371  My Favorite Wife  1940  Cary Grant  actor         Nick  2.0
840381    Penny Serenade  1941  Cary Grant  actor  Roger Adams  2.0

### List the leading roles that Cary Grant played in the 1940s in order by year.

In [127]:
cast[(cast.name=="Cary Grant")&(cast.year//10==194)&(cast.n==1)].sort_values(by='year')

title  year        name   type  \
840397               The Howards of Virginia  1940  Cary Grant  actor   
840353                       His Girl Friday  1940  Cary Grant  actor   
840399                The Philadelphia Story  1940  Cary Grant  actor   
840386                             Suspicion  1941  Cary Grant  actor   
840401                  The Talk of the Town  1942  Cary Grant  actor   
840377                 Once Upon a Honeymoon  1942  Cary Grant  actor   
840344                     Destination Tokyo  1943  Cary Grant  actor   
840369                             Mr. Lucky  1943  Cary Grant  actor   
840370                             Mr. Lucky  1943  Cary Grant  actor   
840378                      Once Upon a Time  1944  Cary Grant  actor   
840336                  Arsenic and Old Lace  1944  Cary Grant  actor   
840373             None But the Lonely Heart  1944  Cary Grant  actor   
840372                         Night and Day  1946  Cary Grant  actor   
840375                             Notorious  1946  Cary Grant  actor   
840393      The Bachelor and the Bobby-Soxer  1947  Cary Grant  actor   
840394                     The Bishop's Wife  1947  Cary Grant  actor   
840368  Mr. Blandings Builds His Dream House  1948  Cary Grant  actor   
840348          Every Girl Should Be Married  1948  Cary Grant  actor   
840357                I Was a Male War Bride  1949  Cary Grant  actor   

                    character    n  
840397            Matt Howard  1.0  
840353           Walter Burns  1.0  
840399     C. K. Dexter Haven  1.0  
840386                Johnnie  1.0  
840401           Leopold Dilg  1.0  
840377  Patrick 'Pat' O'Toole  1.0  
840344          Capt. Cassidy  1.0  
840369              Joe Adams  1.0  
840370        Joe Bascopolous  1.0  
840378            Jerry Flynn  1.0  
840336      Mortimer Brewster  1.0  
840373             Ernie Mott  1.0  
840372            Cole Porter  1.0  
840375                 Devlin  1.0  
840393            Dick Nugent  1.0  
840394                 Dudley  1.0  
840368          Jim Blandings  1.0  
840348      Dr. Madison Brown  1.0  
840357  Captain Henri Rochard  1.0

### How many roles were available for actors in the 1950s?

In [135]:
len(cast[(cast.year // 10 == 195)&(cast.type == "actor")])

154727

#### 1,54,727 roles was available for actors in the 1950s.

### How many roles were available for actresses in the 1950s?

In [141]:
len(cast[(cast.year //10 ==195)&(cast.type=="actress")])

56259

#### 56,259 actress in the year 1950s

### How many leading roles (n=1) were available from the beginning of film history through 1980?

In [144]:
len(cast[(cast.year<=1980)&(cast.n==1)])

63924

#### there is 63,924 leading roles available from the beginning of film history through 1980.

### How many non-leading roles were available through from the beginning of film history through 1980?

In [146]:
len(cast[(cast.year <= 1980) & (cast.n != 1)])

1094780

#### 10,94,780 non - leading roles were avaible through from the beginning of film history through 1980.

### How many roles through 1980 were minor enough that they did not warrant a numeric "n" rank?

In [152]:
len(cast[(cast.year<=1980)&(cast.n.isnull())])

438086

#### 4,38,086 roles through 1980 were minor enough that they did not warrant a numeric "n" rank.